// @stoplight/prism-http/dist/mocker/generator/JSONSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateStatic = exports.sortSchemaAlphabetically = exports.generate = exports.resetGenerator = void 0;
const faker_1 = require("@faker-js/faker");
const lodash_1 = require("lodash");
const JSONSchemaFaker = require("json-schema-faker");
const sampler = require("@stoplight/json-schema-sampler");
const Either_1 = require("fp-ts/Either");
const function_1 = require("fp-ts/function");
const E = require("fp-ts/lib/Either");
const filterRequiredProperties_1 = require("../../utils/filterRequiredProperties");
JSONSchemaFaker.extend('faker', () => faker_1.default);
const JSON_SCHEMA_FAKER_DEFAULT_OPTIONS = Object.fromEntries([
    ['defaultInvalidTypeProduct', null],
    ['defaultRandExpMax', 10],
    ['pruneProperties', []],
    ['ignoreProperties', []],
    ['ignoreMissingRefs', false],
    ['failOnInvalidTypes', true],
    ['failOnInvalidFormat', true],
    ['alwaysFakeOptionals', false],
    ['optionalsProbability', false],
    ['fixedProbabilities', false],
    ['useExamplesValue', false],
    ['useDefaultValue', false],
    ['requiredOnly', false],
    ['minItems', 0],
    ['maxItems', null],
    ['minLength', 0],
    ['maxLength', null],
    ['refDepthMin', 0],
    ['refDepthMax', 3],
    ['resolveJsonPath', false],
    ['reuseProperties', false],
    ['sortProperties', null],
    ['fillProperties', true],
    ['random', Math.random],
    ['replaceEmptyByRandomValue', false],
    ['omitNulls', false],
]);
function resetGenerator() {
    JSONSchemaFaker.option({
        ...JSON_SCHEMA_FAKER_DEFAULT_OPTIONS,
        failOnInvalidTypes: false,
        failOnInvalidFormat: false,
        alwaysFakeOptionals: true,
        optionalsProbability: 1,
        fixedProbabilities: true,
        ignoreMissingRefs: true,
    });
}
exports.resetGenerator = resetGenerator;
resetGenerator();
function generate(bundle, source) {
    return (0, function_1.pipe)((0, filterRequiredProperties_1.stripWriteOnlyProperties)(source), E.fromOption(() => Error('Cannot strip writeOnly properties')), E.chain(updatedSource => (0, Either_1.tryCatch)(() => sortSchemaAlphabetically(JSONSchemaFaker.generate({ ...(0, lodash_1.cloneDeep)(updatedSource), __bundled__: bundle })), Either_1.toError)));
}
exports.generate = generate;
function sortSchemaAlphabetically(source) {
    if (source && Array.isArray(source)) {
        for (const i of source) {
            if (typeof source[i] === 'object') {
                source[i] = sortSchemaAlphabetically(source[i]);
            }
        }
        return source;
    }
    else if (source && typeof source === 'object') {
        Object.keys(source).forEach((key) => {
            if (typeof source[key] === 'object') {
                source[key] = sortSchemaAlphabetically(source[key]);
            }
        });
        return Object.fromEntries(Object.entries(source).sort());
    }
    return source;
}
exports.sortSchemaAlphabetically = sortSchemaAlphabetically;
function generateStatic(resource, source) {
    return (0, Either_1.tryCatch)(() => sampler.sample(source, {}, resource), Either_1.toError);
}
exports.generateStatic = generateStatic;
